REGRESSION (251714@main): Video playback pauses when invoking PiP

youssefsoli opened the pull request:
#### 7607cc59c2c1382d37e10b4dae7311b1e0bbb7d5
<pre>
REGRESSION (251714@main): Video playback pauses when invoking PiP
<a href="https://bugs.webkit.org/show_bug.cgi?id=242922">https://bugs.webkit.org/show_bug.cgi?id=242922</a>
&lt;rdar://96657600&gt;

Reviewed by Wenson Hsieh and Eric Carlson.

This regression was caused because the original logic fix treated regular
fullscreen and PiP as the same &quot;going into fullscreen&quot; action when it should
have only reacted to normal fullscreen rather than PiP.

To fix this issue and the issue of PiP not being able to be entered while
in fullscreen, we can just check the fullscreen mode in canEnterVideoFullscreen
and ignore anything that isn&apos;t normal fullscreen.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::canEnterVideoFullscreen const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::canEnterVideoFullscreen const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::canEnterVideoFullscreen const):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::canEnterVideoFullscreen const):

Canonical link: <a href="https://commits.webkit.org/252666@main">https://commits.webkit.org/252666@main</a>
</pre>

whsieh reviewed (approved):
Seems reasonable to me!

(Adding Jer and Eric, too)

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6402,7 +6402,7 @@ void HTMLMediaElement::enterFullscreen(VideoFullscreenMode mode)
             HTMLVideoElement& asVideo = downcast<HTMLVideoElement>(*this);
             auto& client = document().page()->chrome().client();
-            if (client.supportsVideoFullscreen(mode) && client.canEnterVideoFullscreen()) {
+            if (client.supportsVideoFullscreen(mode) && client.canEnterVideoFullscreen(mode)) {
```
Nit: it would be slightly more efficient to put `client.supportsVideoFullscreen(mode)` into a variable

eric-carlson reviewed (approved):
r=me

youssefsoli committed:
REGRESSION (251714@main): Video playback pauses when invoking PiP
https://bugs.webkit.org/show_bug.cgi?id=242922
<rdar://96657600>

Reviewed by Wenson Hsieh and Eric Carlson.

This regression was caused because the original logic fix treated regular
fullscreen and PiP as the same "going into fullscreen" action when it should
have only reacted to normal fullscreen rather than PiP.

To fix this issue and the issue of PiP not being able to be entered while
in fullscreen, we can just check the fullscreen mode in canEnterVideoFullscreen
and ignore anything that isn't normal fullscreen.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::enterFullscreen):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::canEnterVideoFullscreen const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::canEnterVideoFullscreen const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.h:
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::canEnterVideoFullscreen const):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::canEnterVideoFullscreen const):

Canonical link: https://commits.webkit.org/252666@main
